File: application/app/domain/model/filters/WhitelistIps.scala
package domain.model.filters

import common.validation.constraints.TraversableSize
import domain.model.{Filter, SetValue}
import javax.validation.constraints.NotNull
import play.mvc.Http.HeaderNames.X_FORWARDED_FOR

import scala.annotation.meta.field

case class WhitelistIps(
                         @(NotNull @field) @(TraversableSize @field)(max = 255) whitelist: Set[String]
                       ) extends Filter {
  override val name: String = Filter.WHITELIST_IPS
  override val value: SetValue = SetValue(whitelist)
  override val filter: Map[String, String] => Boolean = {
    headers => headers.contains(X_FORWARDED_FOR) && whitelist.contains(headers(X_FORWARDED_FOR))
  }
}

File: ws-gateway-domain/src/main/java/com/cosmin/wsgateway/domain/model/filters/WhitelistIps.java
package com.cosmin.wsgateway.domain.model.filters;

import com.cosmin.wsgateway.domain.model.Filter;

import lombok.Value;

import java.util.Map;
import java.util.Set;

@Value
public class WhitelistIps implements Filter<Set<String>> {
    private final Set<String> hosts;

    @Override
    public String name() {
        return "whitelistIps";
    }

    @Override
    public Set<String> value() {
        return hosts;
    }

    @Override
    public boolean isAllowed(Map<String, String> requestHeaders) {
        return !requestHeaders.containsKey("Host") && hosts.contains(requestHeaders.get("Host"));
    }
}
